PotionEffect Retrieval Bug

**Sponge API:** 6.0.0-20170323.180311-60
**Sponge Forge Build:** spongeforge-1.11.2-2227-6.0.0-BETA-2244
**Forge Build:** forge-1.11.2-13.20.0.2227-universal
**Java Version:** 1.8.0_121-b13

I've seem to have found a bug, which simon816 has confirmed on the forums, with the with `DataSerializable` in regards to `PotiotionEffect`s. 

**Here is the original post:**

I have the following code, which saves a player's potion effects:
```
public static void saveOriginalPotionEffects(List<PotionEffect> potionEffects, String arenaName, String userName) throws ObjectMappingException {

	int i = 1;
	
	for(PotionEffect potionEffect: potionEffects){
		
		UnversalConfigs.getConfig(ArenaConfig).getNode("Arena", arenaName, "Contestants", userName, "Potion Effects","Effect "+i).setValue(TypeToken.of(PotionEffect.class),potionEffect);
		
		UnversalConfigs.saveConfig(ArenaConfig);
		
		i++;
	}
}
```

I have no issue with the code above, but when I try to retrieve the data with the code below, I receive a `Could not translate DataSerializable of type: org.spongepowered.api.effect.potion.PotionEffect` error. This might be a programmer end error, but I coudn't get any further after troubleshooting the issue to line `PotionEffect potionEffect = node.getNode("Effect "+i).getValue(TypeToken.of(PotionEffect.class)); `Any help would be greatly appreciated. Thanks!

```
public static List<PotionEffect> fetchOriginalPotionEffects(String arenaName, String userName) throws ObjectMappingException{
	
	List<PotionEffect> potionEffects = Lists.newArrayList();
	
	CommentedConfigurationNode node = UnversalConfigs.getConfig(ArenaConfig).getNode("Arena", arenaName, "Contestants", userName, "Potion Effects");
	
	int NOPE = getNumOfPotionEffects(arenaName,userName);
	
	for(int i = 1; i<= NOPE;i++){
		
		PotionEffect potionEffect = node.getNode("Effect "+i).getValue(TypeToken.of(PotionEffect.class));
		
		potionEffects.add(potionEffect);

	}
	
	return potionEffects;
}

```
**simon816's bug replication code:**

```
try {
    PotionEffect effect = PotionEffect.of(PotionEffectTypes.JUMP_BOOST, 1, 10);
    SimpleCommentedConfigurationNode node = SimpleCommentedConfigurationNode.root();
    node.getNode("test").setValue(TypeToken.of(PotionEffect.class), effect);
    PotionEffect effect2 = node.getNode("test").getValue(TypeToken.of(PotionEffect.class));
} catch (ObjectMappingException e) {
    e.printStackTrace();
}
```

I think I accidentally managed to find the core issue.

When trying to to get a PotionEffect from a DataContainer with this code:

```
PotionEffect effect = PotionEffect.of(PotionEffectTypes.ABSORPTION, 0, 100);
PotionEffect.builder().build(effect.toContainer());
```

I get this error:

```
[17:11:12 ERROR] [STDERR]: org.spongepowered.api.data.persistence.InvalidDataException: Could not deserialize something correctly, likely due to bad type data.
[17:11:12 ERROR] [STDERR]:      at org.spongepowered.api.data.persistence.AbstractDataBuilder.build(AbstractDataBuilder.java:89)
[17:11:12 ERROR] [STDERR]:      at test.code.Main.main(Main.java:2)
[17:11:13 ERROR] [STDERR]: Caused by: org.spongepowered.api.data.persistence.InvalidDataException: The container has an invalid potion type name: effect.absorption
[17:11:13 ERROR] [STDERR]:      at org.spongepowered.common.effect.potion.SpongePotionBuilder.buildContent(SpongePotionBuilder.java:75)
[17:11:13 ERROR] [STDERR]:      at org.spongepowered.api.data.persistence.AbstractDataBuilder.build(AbstractDataBuilder.java:87)
[17:11:13 ERROR] [STDERR]:      ... 2 more
```
*(Stacktrace shortened and simplified)*

Not sure what format SpongePotionBuilder expects but it's not `effect.absorbtion`.

--- a/src/main/java/org/spongepowered/common/data/DataRegistrar.java
+++ b/src/main/java/org/spongepowered/common/data/DataRegistrar.java
@@ -41,6 +41,7 @@ import org.spongepowered.api.data.manipulator.mutable.block.*;
 import org.spongepowered.api.data.manipulator.mutable.entity.*;
 import org.spongepowered.api.data.manipulator.mutable.item.*;
 import org.spongepowered.api.data.manipulator.mutable.tileentity.*;
+import org.spongepowered.api.effect.potion.PotionEffect;
 import org.spongepowered.api.item.enchantment.Enchantment;
 import org.spongepowered.api.data.meta.PatternLayer;
 import org.spongepowered.api.data.property.PropertyRegistry;
@@ -73,6 +74,7 @@ import org.spongepowered.common.data.builder.item.*;
 import org.spongepowered.common.data.builder.manipulator.InvisibilityDataAddVanishUpdater;
 import org.spongepowered.common.data.builder.manipulator.immutable.block.ImmutableSpongeTreeDataBuilder;
 import org.spongepowered.common.data.builder.manipulator.immutable.item.ImmutableItemEnchantmentDataBuilder;
+import org.spongepowered.common.effect.potion.PotionEffectContentUpdater;
 import org.spongepowered.common.item.enchantment.SpongeEnchantmentBuilder;
 import org.spongepowered.common.data.builder.meta.SpongePatternLayerBuilder;
 import org.spongepowered.common.data.builder.util.weighted.BaseAndAdditionBuilder;
@@ -188,6 +190,12 @@ public class DataRegistrar {
         dataManager.registerContentUpdater(ImmutableInvisibilityData.class, invisibilityUpdater);
         dataManager.registerContentUpdater(SpongeInvisibilityData.class, invisibilityUpdater);
         dataManager.registerContentUpdater(ImmutableSpongeInvisibilityData.class, invisibilityUpdater);
+        final PotionEffectContentUpdater potionUpdater = new PotionEffectContentUpdater();
+        dataManager.registerContentUpdater(PotionEffect.class, potionUpdater);
+        dataManager.registerContentUpdater(PotionEffectData.class, potionUpdater);
+        dataManager.registerContentUpdater(ImmutablePotionEffectData.class, potionUpdater);
+        dataManager.registerContentUpdater(SpongePotionEffectData.class, potionUpdater);
+        dataManager.registerContentUpdater(ImmutableSpongePotionEffectData.class, potionUpdater);
 
         // Content Updaters for Custom Data
         dataManager.registerContentUpdater(DataManipulator.class, new LegacyCustomDataClassContentUpdater());

--- a/src/main/java/org/spongepowered/common/data/util/DataVersions.java
+++ b/src/main/java/org/spongepowered/common/data/util/DataVersions.java
@@ -93,6 +93,17 @@ public final class DataVersions {
 
     }
 
+    public static final class Potion {
+
+        public static final int BROKEN_POTION_ID = 1;
+        public static final int POTION_V2 = 2;
+        public static final int CURRENT_VERSION = POTION_V2;
+
+        private Potion() {
+        }
+
+    }
+
     public static final class Data {
 
         public static final int INVISIBILITY_DATA_PRE_1_9 = 1;

--- a/src/main/java/org/spongepowered/common/effect/potion/PotionEffectContentUpdater.java
+++ b/src/main/java/org/spongepowered/common/effect/potion/PotionEffectContentUpdater.java
@@ -1,0 +1,62 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.effect.potion;
+
+import com.google.common.collect.ImmutableMap;
+import org.spongepowered.api.CatalogType;
+import org.spongepowered.api.data.DataView;
+import org.spongepowered.api.data.persistence.DataContentUpdater;
+import org.spongepowered.common.data.util.DataQueries;
+import org.spongepowered.common.data.util.DataVersions;
+import org.spongepowered.common.registry.type.effect.PotionEffectTypeRegistryModule;
+
+import java.util.stream.Collectors;
+
+public class PotionEffectContentUpdater implements DataContentUpdater {
+
+    private final static ImmutableMap<String, String> map = ImmutableMap.<String, String>builder().putAll(PotionEffectTypeRegistryModule.getInstance().getAll()
+            .stream().collect(Collectors.toMap(CatalogType::getName, CatalogType::getId))).build();
+
+    @Override
+    public int getInputVersion() {
+        return DataVersions.Potion.BROKEN_POTION_ID;
+    }
+
+    @Override
+    public int getOutputVersion() {
+        return DataVersions.Potion.POTION_V2;
+    }
+
+    @Override
+    public DataView update(DataView content) {
+        final String oldId = content.getString(DataQueries.POTION_TYPE).get();
+        final String newId = map.get(oldId);
+
+        content.remove(DataQueries.POTION_TYPE);
+        content.set(DataQueries.POTION_TYPE, newId);
+
+        return content;
+    }
+}

--- a/src/main/java/org/spongepowered/common/effect/potion/SpongePotionBuilder.java
+++ b/src/main/java/org/spongepowered/common/effect/potion/SpongePotionBuilder.java
@@ -36,6 +36,7 @@ import org.spongepowered.api.effect.potion.PotionEffect;
 import org.spongepowered.api.effect.potion.PotionEffectType;
 import org.spongepowered.api.data.persistence.AbstractDataBuilder;
 import org.spongepowered.common.data.util.DataQueries;
+import org.spongepowered.common.data.util.DataVersions;
 
 import java.util.Optional;
 
@@ -48,7 +49,7 @@ public class SpongePotionBuilder extends AbstractDataBuilder<PotionEffect> imple
     private boolean showParticles;
 
     public SpongePotionBuilder() {
-        super(PotionEffect.class, 1);
+        super(PotionEffect.class, DataVersions.Potion.CURRENT_VERSION);
     }
 
     @Override

--- a/src/main/java/org/spongepowered/common/mixin/core/potion/MixinPotionEffect.java
+++ b/src/main/java/org/spongepowered/common/mixin/core/potion/MixinPotionEffect.java
@@ -37,6 +37,7 @@ import org.spongepowered.asm.mixin.Intrinsic;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.Shadow;
 import org.spongepowered.common.data.util.DataQueries;
+import org.spongepowered.common.data.util.DataVersions;
 
 @NonnullByDefault
 @Mixin(net.minecraft.potion.PotionEffect.class)
@@ -76,14 +77,14 @@ public abstract class MixinPotionEffect implements PotionEffect {
 
     @Override
     public int getContentVersion() {
-        return 1;
+        return DataVersions.Potion.CURRENT_VERSION;
     }
 
     @Override
     public DataContainer toContainer() {
         return DataContainer.createNew()
                 .set(Queries.CONTENT_VERSION, getContentVersion())
-                .set(DataQueries.POTION_TYPE, this.potion.getName())
+                .set(DataQueries.POTION_TYPE, this.getType().getId())
                 .set(DataQueries.POTION_DURATION, this.duration)
                 .set(DataQueries.POTION_AMPLIFIER, this.amplifier)
                 .set(DataQueries.POTION_AMBIANCE, this.isAmbient)

